tests/components/template/test_vacuum.py

```diff
@@ -0,0 +1,473 @@
+
+    _verify(hass, STATE_UNKNOWN, None)
+
+
+async def test_invalid_templates(hass, calls):
+    """Test invalid templates."""
+    with assert_setup_component(1, 'vacuum'):
+        assert await setup.async_setup_component(hass, 'vacuum', {
+            'vacuum': {
+                'platform': 'template',
+                'vacuums': {
+                    'test_vacuum': {
+                        'value_template':
+                            "{{ this_function_does_not_exist() }}",
+                        'battery_level_template':
+                            "{{ this_function_does_not_exist() }}",
+                        'fan_speed_template':
+                            "{{ this_function_does_not_exist() }}",
+                        'start': {
+                            'service': 'script.vacuum_start'
+                        }
+                    }
+                }
+            }
+        })
+
+    await hass.async_start()
+    await hass.async_block_till_done()
+
+    _verify(hass, STATE_UNKNOWN, None)
+
+# End of template tests #
+
+
+# Function tests #
+async def test_state_services(hass, calls):
+    """Test state services."""
+    await _register_components(hass)
+
+    # Start vacuum
+    common.async_start(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # verify
+    assert hass.states.get(_STATE_INPUT_SELECT).state == STATE_CLEANING
+    _verify(hass, STATE_CLEANING, None)
+
+    # Pause vacuum
+    common.async_pause(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # verify
+    assert hass.states.get(_STATE_INPUT_SELECT).state == STATE_PAUSED
+    _verify(hass, STATE_PAUSED, None)
+
+    # Stop vacuum
+    common.async_stop(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # verify
+    assert hass.states.get(_STATE_INPUT_SELECT).state == STATE_IDLE
+    _verify(hass, STATE_IDLE, None)
+
+    # Return vacuum to base
+    common.async_return_to_base(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # verify
+    assert hass.states.get(_STATE_INPUT_SELECT).state == STATE_RETURNING
+    _verify(hass, STATE_RETURNING, None)
+
+
+async def test_unused_services(hass, calls):
+    """Test calling unused services should not crash."""
+    await _register_basic_vacuum(hass)
+
+    # Pause vacuum
+    common.async_pause(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # Stop vacuum
+    common.async_stop(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # Return vacuum to base
+    common.async_return_to_base(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # Spot cleaning
+    common.async_clean_spot(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # Locate vacuum
+    common.async_locate(hass, _TEST_VACUUM)
+    await hass.async_block_till_done()
+
+    # Set fan's speed
+    common.async_set_fan_speed(hass, 'medium', _TEST_VACUUM)
+    await hass.async_block_till_done()
+
```

Comment:
Nothing is asserted here. Each test must assert something.

Reply:
This test is ment to test of the services correctly handle the situations where no services is configured (the `script is None`-check). What is the best way to assert that in a unit test? The current version works, because the script-call would cause an exception when no script is provided. But there's probably a neater way that I don't know off yet 😄.

I've also added an assert to verify if the state is still on the initial value. The state shouldn't be changed by the component when an unconfigured service is called.

Reply:
I think it's ok like this.